docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+# Kubernetes Code Generators
+
+Kubernetes provides a collection of code generators to assist with developing APIs.  These code generators generate
+useful libraries such as clients, openapi definitions, copying go structs, etc.
+
+The `kubegen` uses directory structure conventions to detect Kubernetes APIs definitions and run the standard
+set of Kubernetes code generators.
+
+There are 2 methods for running kubegen
+
+- Directly through the kubegen command line by downloading the binary and running at a the project root
+- Through Bazel by adding rules to `WORKSPACE` and `BUILD.gazel`
+
+## Kubernetes API conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups defined under `pkg/apis/<group>/<version>` where *group* and *version* match the following
+patterns.
+
+- **group pattern**: ^[a-z]+$
+- **version pattern**: ^v\\d+(alpha\\d+|beta\\d+)*$
+
+By default, kubegen will run code generators for both external types defined under `pkg/apis/<group>/<version>` and
+internal types defined under `pkg/apis/<group>`.  The location kubegen searches can be overridden.
+
+kubegen will prepend all generated files with copyright and license comments.
+
+By default, the kubegen binary runs all* of the Kubernetes code
+generators against all API group/versions defined in the project.
+
+**List of generators:*
+
+- client-gen
+- conversion-gen
+- deepcopy-gen
+- defaulter-gen
+- go-to-protobuf
+- informer-gen
+- lister-gen
+- openapi-gen
+- set-gen
```

Comment:
set-gen is independent. It's not run by API groups.

Reply:
done